torch/ao/quantization/dynamic_tracing/quantization_state.py

```diff
@@ -0,0 +1,797 @@
+                # TODO: handle non-tensor inputs
+                # TODO: this is not handling non-tensor tuple args (for example,
+                # dilation in conv2d) correctly, it just happens to work but
+                # needs a fix.
+                quant_info = arg_quant_infos[tensor_arg_idx]
+                dequant_info = arg_dequant_infos[tensor_arg_idx]
+                if quant_info is not None:
+                    scale, zp = quant_info
+                    arg = torch.quantize_per_tensor(arg, scale, zp, torch.quint8)
+                elif dequant_info is True:
+                    arg = arg.dequantize()
+                new_args.append(arg)
+                tensor_arg_idx += 1
+
+        # if there is a packed param, replace the relevant args
+        if packed_param_name is not None:
+            new_args_with_packed = []
+            packable_arg_idxs = get_packable_arg_idxs(orig_op)
+            added_packed = False
+            for idx, arg in enumerate(new_args):
+                if packable_arg_idxs is not None and idx in packable_arg_idxs:
+                    if not added_packed:
+                        packed_param = getattr(root_module, packed_param_name)
+                        new_args_with_packed.append(packed_param)
+                        added_packed = True
+                else:
+                    new_args_with_packed.append(arg)
+            new_args = new_args_with_packed
+
+        # potentially extend kwargs with scale and zero_point
+        # TODO move op-specific logic out of here
+        if orig_op not in (F.conv2d, F.linear):
+            kwargs.update(**additional_kwargs)
+        else:
+            seen_op = self._get_cur_seen_op()
+            if seen_op.output_tensor_infos[0].inf_dtype == torch.quint8:
+                new_args.append(additional_kwargs['scale'])
+                new_args.append(additional_kwargs['zero_point'])
+
+        # TODO move op-specific logic out of here
+        if op is torch.ops.quantized.linear:
+            del kwargs['bias']
+
+        return op, tuple(new_args), kwargs
+
+    def op_convert_after_hook(
+        self,
+        op: Callable,
+        output,
+    ) -> Any:
+        """
+        This function is called aftern an op call in a converted model.
+
+        TODO: add dequant, if needed
+        """
+        return output
+
+    def get_op_convert_info(
+        self,
+        op: Callable,
+        unwrap_scale_zp: bool = False,
+    ) -> Tuple[
+        Callable,
+        List[Optional[Tuple[float, int]]],
+        List[bool],
+        Optional[str],
+        Dict[str, Any]
+    ]:
+        """
+        Returns the information needed for convert time modifications to `op`.
+        Has no side effects.  Return types:
+
+        `new_op`. Returns either the original callable unchanged,
+        or the corresponding quantized target. Note: always returns the original
+        callable for modules, because they are quantized with module swaps.
```

Comment:
is this quantized with torch.ao.quantization.convert? wondering if we should incorporate the swapping here as well and remove the dependency on that?

Reply:
This whole function is a no-op for modules, because they are using `convert`.  We could consider a refactor to not use `convert`, in that case maybe it would make sense to not use `prepare` as well.